### PR TITLE
8288883: C2: assert(allow_address || t != T_ADDRESS) failed after JDK-8283091

### PR DESCRIPTION
**Pull request by fg1417**
Superword doesn't vectorize any nodes of non-primitive types and
thus sets `allow_address` false when calling type2aelembytes() in
SuperWord::data_size()[1]. Therefore, when we try to resolve the
data size for a node of T_ADDRESS type, the assertion in
type2aelembytes()[2] takes effect.

We try to resolve the data sizes for node s and node t in the
SuperWord::adjust_alignment_for_type_conversion()[3] when type
conversion between different data sizes happens. The issue is,
when node s is a ConvI2L node and node t is an AddP node of
T_ADDRESS type, type2aelembytes() will assert. To fix it, we
should filter out all non-primitive nodes, like the patch does
in SuperWord::adjust_alignment_for_type_conversion(). Since
it's a failure in the mid-end, all superword available platforms
are affected. In my local test, this failure can be reproduced
on both x86 and aarch64. With this patch, the failure can be fixed.

Apart from fixing the bug, the patch also adds necessary type check
and does some clean-up in SuperWord::longer_type_for_conversion()
and VectorCastNode::implemented().

[1]https://github.com/openjdk/jdk/blob/dddd4e7c81fccd82b0fd37ea4583ce1a8e175919/src/hotspot/share/opto/superword.cpp#L1417
[2]https://github.com/openjdk/jdk/blob/b96ba19807845739b36274efb168dd048db819a3/src/hotspot/share/utilities/globalDefinitions.cpp#L326
[3]https://github.com/openjdk/jdk/blob/dddd4e7c81fccd82b0fd37ea4583ce1a8e175919/src/hotspot/share/opto/superword.cpp#L1454

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288883](https://bugs.openjdk.org/browse/JDK-8288883): C2: assert(allow_address || t != T_ADDRESS) failed after JDK-8283091

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9391/head:pull/9391` \
`$ git checkout pull/9391`

Update a local copy of the PR: \
`$ git checkout pull/9391` \
`$ git pull https://git.openjdk.org/jdk pull/9391/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9391`

View PR using the GUI difftool: \
`$ git pr show -t 9391`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9391.diff">https://git.openjdk.org/jdk/pull/9391.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
In which call to `adjust_alignment_for_type_conversion()` you got AddP node?
Should we add checks there too?

**Comment by fg1417** [comment]
> In which call to `adjust_alignment_for_type_conversion()` you got AddP node? Should we add checks there too?

Thanks for your review, @vnkozlov .

When we call `adjust_alignment_for_type_conversion()` in `SuperWord::follow_def_uses()`, https://github.com/openjdk/jdk/blob/3f1174aa4709aabcfde8b40deec88b8ed466cc06/src/hotspot/share/opto/superword.cpp#L1525, we got AddP node. In this function, we also call `stmts_can_pack()` on the next line, which has checks to prevent unwanted pairs, https://github.com/openjdk/jdk/blob/3f1174aa4709aabcfde8b40deec88b8ed466cc06/src/hotspot/share/opto/superword.cpp#L1202. Maybe we don't have to add one more. WDYT?

**Comment by vnkozlov** [comment]
@fg1417   `stmts_can_pack()` is called in an other place which is preceded by `are_adjacent_refs()` call which also has primitive type check (but different). I was thinking to convert checks in `stmts_can_pack()` to `assert`. But, on other hand, `is_java_primitive(bt)` is cheap and I would prefer to keep checks in `stmts_can_pack()` as they are in case we call it in an other place.

Anyway. After looking on code I agree with your current changes. Let me test it. And you need second review.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

**Comment by fg1417** [comment]
Thanks for your review and test work, @vnkozlov .

May I have a second review please?

**Comment by fg1417** [comment]
> Your fix LGTM. The test doesn't show the problem on PPC64, but my original replay file has worked to verify the fix.

Thanks for your review and verification, @TheRealMDoerr .

**Pull request closed without merging**
